### ISSUE TITLE
setting Subcatchment Parameters for calibration purposes

### ISSUE THREAD

**Issue by bjamali** [OpenWaterAnalytics/pyswmm#246]
It seems that Subcatchment parameters that can be set by user are limited to width,  area,  fracImperv, slope  and curbLength (see class SubcParams(Enum) in toolkitapi.py). 

For calibration of  a SWMM model it is usually necessary to change parameters such as previous and impervious roughness, infiltration rates, depression storage, etc. Is this functionality already available?

**Comment by jennwuu** [comment]
Are you looking to change these parameters before simulation starts?

**Comment by bjamali** [comment]
Yes, thats correct.

**Comment by jennwuu** [comment]
I don't think this feature is available in the SWMM toolkitapi, which is what pyswmm use. I would make the request here: https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/issues

**Comment by yangdong963** [comment]
Hello, I am studying pyswmm recently, and I have a question to ask you. The parameter of changing the impervious area of the subcatchment cannot be modified. Is it my code problem? ‘’‘S1.percent_impervious = 10’

**Comment by yangdong963** [comment]
code:print(s1.percent_impervious)
    s1.percent_impervious = 10
    print(s1.percent_impervious)
output:
0.29
0.29

**Comment by katmratliff** [comment]
Hi @yangdong963, can you please post a larger portion of your code? Thanks!

**Comment by yangdong963** [comment]
from pyswmm import Simulation, Subcatchments
with Simulation('./C1C2.inp') as sim:
    s1 = Subcatchments(sim)['S1']
    s1.slope = 100
    s1.percent_impervious = 0
    s1.percent_impervious = 0.22
    print(sub.percent_impervious)




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "OpenWaterAnalytics/pyswmm"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年7月22日(星期三) 晚上8:41
收件人:&nbsp;"OpenWaterAnalytics/pyswmm"<pyswmm@noreply.github.com&gt;;
抄送:&nbsp;"杨先森"<841003445@qq.com&gt;;"Mention"<mention@noreply.github.com&gt;;
主题:&nbsp;Re: [OpenWaterAnalytics/pyswmm] setting Subcatchment Parameters for calibration purposes (#246)





 
Hi @yangdong963, can you please post a larger portion of your code? Thanks!

**Comment by yangdong963** [comment]
from pyswmm import Simulation, Subcatchments
with Simulation('./C1C2.inp') as sim:
    s1 = Subcatchments(sim)['S1']
    s1.slope = 100
    print(s1.percent_impervious)
    s1.percent_impervious = 0.22
    print(s1.percent_impervious)
Hello, when I tried to modify the percent_impervious parameter in subcatchments, I found that this parameter cannot be modified, and parameters such as slope can be modified. My python environment is python 3.7. Thank you. Looking forward to your reply.
------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "OpenWaterAnalytics/pyswmm"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年7月22日(星期三) 晚上8:41
收件人:&nbsp;"OpenWaterAnalytics/pyswmm"<pyswmm@noreply.github.com&gt;;
抄送:&nbsp;"杨先森"<841003445@qq.com&gt;;"Mention"<mention@noreply.github.com&gt;;
主题:&nbsp;Re: [OpenWaterAnalytics/pyswmm] setting Subcatchment Parameters for calibration purposes (#246)





 
Hi @yangdong963, can you please post a larger portion of your code? Thanks!

**Comment by katmratliff** [comment]
thanks @yangdong963. As @bemcdonnell pointed out [here](https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/issues/298#issuecomment-662435305), this feature is not yet available. From the comments in the code, it looks like this may be a more challenging setter to implement, as it will require recalculation of other parameters after setting. I believe @jennwuu is working on some of the subcatchment parameters, but I'm not sure about this one (?)

**Comment by yangdong963** [comment]
Thank you for answering my question. In the process of calibrating the model, the parameters in subcatchments such as N-Imperv, N-perv, Dstore-imperv and other parameters are also empirical parameters that can be adjusted. I don’t know if you will consider them in the next development process. Yet?
One last word, thank you




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "OpenWaterAnalytics/pyswmm"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年7月22日(星期三) 晚上9:05
收件人:&nbsp;"OpenWaterAnalytics/pyswmm"<pyswmm@noreply.github.com&gt;;
抄送:&nbsp;"杨先森"<841003445@qq.com&gt;;"Mention"<mention@noreply.github.com&gt;;
主题:&nbsp;Re: [OpenWaterAnalytics/pyswmm] setting Subcatchment Parameters for calibration purposes (#246)





 
thanks @yangdong963. As @bemcdonnell pointed out here, this feature is not yet available. From the comments in the code, it looks like this may be a more challenging setter to implement, as it will require recalculation of other parameters after setting. I believe @jennwuu is working on some of the subcatchment parameters, but I'm not sure about this one (?)

**Comment by bemcdonnell** [comment]
@yangdong963, would you be interested in contributing these features?

**Comment by yangdong963** [comment]
Thank you, I'm just new to programming and I still don't have the ability to contribute




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "OpenWaterAnalytics/pyswmm"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年7月22日(星期三) 晚上9:18
收件人:&nbsp;"OpenWaterAnalytics/pyswmm"<pyswmm@noreply.github.com&gt;;
抄送:&nbsp;"杨先森"<841003445@qq.com&gt;;"Mention"<mention@noreply.github.com&gt;;
主题:&nbsp;Re: [OpenWaterAnalytics/pyswmm] setting Subcatchment Parameters for calibration purposes (#246)





 
@yangdong963, would you be interested in contributing these features?

**Comment by jennwuu** [comment]
@yangdong963 yes, N-Imperv, N-perv, Dstore-imperv, and Dstore-perv are parameters I am planning to expand in subcatchment getters and setters.

**Comment by tamaddun** [comment]
@yangdong963 Are you aware of any update on this?

**Comment by jennwuu** [comment]
@tamaddun I am still working on this feature. It's more complex than I originally thought and that's why it is delayed.

**Comment by tamaddun** [comment]
@jennwuu Thank you. Rooting for you!

**Comment by BuczynskiRafal** [comment]
Okay, I see what the problem is. 
The validation in question is planned to be on the PySWMM side? I might be able to implement this? 

```c
                break; //Subcatch[index].fracImperv; break;
                // Cannot Open Function just yet.  Need
                // to adjust some internal functions to
                // ensure parameters are recalculated
                // = MIN(value, 100.0) / 100.0; break;
```

**Comment by bemcdonnell** [comment]
@BuczynskiRafal, if you scroll down to the bottom of this page it will show you the links to the various repositories and how the project is architected. 

https://www.pyswmm.org/ 

I’m going to convert the issue to a discussion until we find units of work to work on
